examples/run_nlp_trainer.py

```diff
@@ -0,0 +1,350 @@
+    
+    args = parser.parse_args()
+    validate_args(args)
+    return args
+
+''' Logic for handling a pretraining job spawn based on storage and training settings.'''
+def run_pretrainer(args):
+    root_dir = os.path.join(os.path.dirname(composer.__file__), '..')
+    training_script = os.path.join(root_dir, 'examples/run_composer_trainer.py')
+    if args.save_locally:
+        args.save_folder = os.path.join(args.run_name, 'checkpoints')
+        if args.training_scheme == 'all':
+            new_save_folder = os.path.join(args.save_folder, 'pretrained')
+        else: # pretrain only
+            new_save_folder = args.save_folder
+        
+        proc = subprocess.Popen(['composer', training_script, '-f', f"{args.file}", '--save_folder', f"{new_save_folder}", 
+                        '--loggers', 'wandb', '--tags', 'pretrain'], cwd=root_dir)
+        
+    # saving on cloud
+    else:
+        if args.training_scheme == 'all':
+            args.save_folder = 'pretrained'
+            save_artifact_name = '{run_name}/pretrained/ep{epoch}-ba{batch}-rank{rank}-wct{total_wct}'
+            proc = subprocess.Popen(['composer', training_script, '-f', f"{args.file}" , '--save_folder', args.save_folder, 
+                    '--save_artifact_name', save_artifact_name, '--save_num_checkpoints_to_keep', '0', '--loggers',
+                    'wandb', 'object_store', '--tags', 'pretrain', '--object_store_hparams', 's3', '--bucket', args.bucket], cwd=root_dir)                
+        
+        # only pretraining, save to default artifact name 
+        else: 
+            proc = subprocess.Popen(['composer', training_script, '-f', f"{args.file}", '--save_num_checkpoints_to_keep', '0',
+                    '--loggers', 'wandb', 'object_store', '--tags', 'pretrain', '--object_store_hparams', 's3', '--bucket', args.bucket], cwd=root_dir) 
+    proc.wait() # block until training is complete 
+
+''' Logic for handling a finetuning job spawn based on storage and training settings.'''
+def run_finetuner(args):
+    # set automatic load and save paths 
+    if args.save_locally:
+        if args.training_scheme == 'all':
+            # load checkpoints from where pretraining saved them
+            ckpt_folder = os.path.join(args.save_folder, 'pretrained')
+            save_folder = args.save_folder
+        else: # finetune only
+            ckpt_folder = args.load_path
+            save_folder = os.path.join(args.load_path, '..')
+
+        all_ckpts_list =  sorted(os.listdir(ckpt_folder))
+            
+    else:
+        if args.training_scheme == 'all':
+            ckpt_folder = f'{args.run_name}/{args.save_folder}'
+            save_folder = args.save_folder
+        else: # finetune only
+            ckpt_folder = args.load_path # load path should be full/path/to/folder/ containing all ckpts 
+            save_folder = args.run_name 
+
+        all_ckpts_list = get_all_s3_checkpoints(args.bucket, ckpt_folder)
```

Comment:
Ah, I think I answer one of my questions from before about why we need to list the bucket. I see why that is needed now.

Instead of requiring a list bucket / list dir, could this function take the list of checkpoints to finetune?